### ISSUE TITLE
spec: state that Starlark behavior is locale-independent

### ISSUE THREAD

**Issue by alandonovan** [bazelbuild/starlark#147]
The behavior of, say, `"%f" % 1.2345` should not be a function of the user's environment. The spec should state that, and the implementations should be audited for operations that are locale-dependent.

See https://github.com/bazelbuild/bazel/pull/12658 for context.

**Comment by brandjon** [comment]
My first thought was that we shouldn't need to specify this at all, it should be assumed. But then again, is formatting specified to be implementation-defined in any way? If so, I don't think we *can* specify that an implementation may not consider the locale.

If formatting is not implementation-defined then I don't know why locale even needs to be mentioned, except to reinforce the point.

**Comment by adonovan** [comment]
> My first thought was that we shouldn't need to specify this at all, it should be assumed. But then again, is formatting specified to be implementation-defined in any way? If so, I don't think we can specify that an implementation may not consider the locale. If formatting is not implementation-defined then I don't know why locale even needs to be mentioned, except to reinforce the point.

The behavior of various C functions such as `printf` and `isalpha` is required by POSIX to honor the [LC_* locale environment variables](https://pubs.opengroup.org/onlinepubs/009695199/basedefs/xbd_chap07.html), so a C implementation of Starlark that depends on `snprintf` for formatting may have a bug. Similarly, Bazel's Java implementation of Starlark uses String.format for formatting, and it observes the LC_* variables (via the Java Locale object) and so has a similar bug.

Perhaps it doesn't need to be stated that Starlark is not C or Java, but the naive implementation strategy behaves like those languages in ways we don't want.

**Comment by brandjon** [comment]
Sure, it's very easy to accidentally have this sensitivity to locale.

The question is whether the spec bans this behavior by fully specifying formatting, or whether the spec permits an implementation to affirmatively choose to depend on locale.

It seems awkward to do something in-between, like allow an implementation to choose a formatting generally, but still mandate that it be consistent across platforms and environment variables. How would that even be interpreted and enforced? It would be like the spec mandating that your implementation is not allowed to support a flag that makes its behavior not compliant with the spec.

**Comment by adonovan** [comment]
Fair enough. Perhaps it's enough to say nothing; but the Java implementation should be fixed.

**Comment by brandjon** [comment]
> it's more important that the form of the output is predictable than that each number is accurate to 1 ulp

If it's not 100% specified then program behavior may vary in arbitrary ways across implementations. But I wouldn't know how to do better. So we can specify what we can and explicitly say implementation-defined for issues like precision.

**Comment by jwnimmer-tri** [comment]
In modern times, the concept of "shortest round-trip prevision" has gained traction, e.g., https://learn.microsoft.com/en-us/cpp/standard-library/charconv-functions?view=msvc-170.  One good implementation is https://github.com/ulfjack/ryu.  For any new languages or APIs, I recommend specifying that as the desired behavior (in the absence of any other formatting modifiers, e.g., if the caller specifically requests truncation).

> it's more important that the form of the output is predictable than that each number is accurate to 1 ulp

I'd agree with that relative priority, but I think that both are must-have requirements.  If I format a float to string and end up with a different number, I would consider that a bug.

**Comment by adonovan** [comment]
Funny, https://github.com/ulfjack/ryu was written by the former TL of Bazel. We don't use it in Starlark, but we're aware that we should: https://github.com/bazelbuild/bazel/blob/5ac37924408db7dfc4e100e01f24b508a7e4b7a6/src/main/java/net/starlark/java/eval/StarlarkFloat.java#L151-L162

P.S. Hey Jeremy, very long time no see! I hope you are well. What brings you to Starlarkland?
